### ISSUE TITLE
Referring Route

### ISSUE THREAD

**Issue by mav2287** [middlewares/ideas#32]
I have some places where I check the referring route. In order to do this I had to create a function that produces the referring route and accounts for the base path. I haven't seen a middleware for this so I figured I would see if it makes sense to make this into a middleware.

```
function getRefererRoute(ServerRequestInterface $request, string $basePath) {
    $httpReferer = $request->getServerParams()["HTTP_REFERER"] ?? false;
    if ($httpReferer == false) {
        return "";
    }

    $parsedURL = parse_url($httpReferer);
    $route = $parsedURL['path'];

    if (substr($route, 0, strlen($basePath)) == $basePath) {
        $route = substr($route, strlen($basePath));
    }

    return $route;
}
```

**Comment by oscarotero** [comment]
I don't think this is a common use case to create a middleware.

**Comment by mav2287** [comment]
My other thought then was if this should be added as an option to the base-path middleware.

**Comment by oscarotero** [comment]
What's the purpose of this? Looking at the code, it get the http-referer header, remove the basepath and return the trailing path. But not sure why you need this.

**Comment by mav2287** [comment]
This is useful to see what route a request came from. So for example you can check that a form submission came from the route where the form exists instead of a programatic submission.

**Comment by oscarotero** [comment]
I don't think this is a good idea. The http-referer header is not safe (it can be changed easily) by the client. A better solution, if you need this value, is insert and send this value in the form (for example, using a hidden input), or add it as a url parameter (like `send?from=referer`).

**Comment by mav2287** [comment]
The only reason I didn't go with that is that it adds an extra parameter and can still be easily changed by the user.

**Comment by oscarotero** [comment]
IMHO, I see some problems:

- A form should include all info needed to perform an action. No matter where it has been submitted. A form that does different things depending on random stuff like a http header is very difficult to test and open the door to many bugs.
- If you don't want the user to change a value, just encode it or use any encrypt method like JSON Web Token, so the user cannot change it, and if he does, the value will be invalid.

**Comment by mav2287** [comment]
I agree that you shouldn't take action based on the referrer and that the form should have everything you need, however there are some common uses cases for example logging where the form was submitted from, preventing forms that were submitted with a referrer that doesn't match and using the referrer to prevent non-malicious hot linking of assets.

**Comment by oscarotero** [comment]
Yes, but in this case, it's better implement a CSRF system because, as said, http headers can be easily modified, for example:

```
$ curl -H "Referer: example.com" -X POST http://my-site.com
```

In PSR-7 middlewares, there is a CSRF middleware (https://github.com/oscarotero/psr7-middlewares#csrf) but it hasn't been ported to PSR-15 because it's too opinionated.

**Comment by mav2287** [comment]
With the form I agree 100% and a CSRF middleware was something I planned to tackle at some point soon. With hot linking it more a case of it being non-malicious so it won't be intentionally spoofed.

With regard to the CSRF implementation is there a library you think makes sense to create a middleware out of? It may just be time for me to put that together.

**Comment by oscarotero** [comment]
The PSR-7 version was inspired by [anti-csrf](https://github.com/paragonie/anti-csrf) library but due it's not PSR-7 compatible, it shouldn't be used to create a middleware.
Maybe the only thing you need is https://github.com/psr7-sessions/storageless, that allows to store data without using $_SESSION but cookies and starting from [v5.0.0](https://github.com/psr7-sessions/storageless/releases/tag/5.0.0) it includes some improvements to prevent CSRF attacks.

**Comment by mav2287** [comment]
I will take a look. I think we may also need to think forward to the fact more and more projects now have JS front ends and PHP backends so any library or middleware setup we go with will need to be able to work within that parameter. I only bring it up because most of the libraries I have worked with in the past inject the field into the form which just isn't possible with a JS front end.
